File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        
        if(!Arrays.asList(words).contains(target)){
            return 0;
        }
        
        Queue <String> q = new LinkedList<>();
        q.offer(begin);
        int answer = 0;
        
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                String current = q.poll();
                
                if(current.equals(target)){
                    return answer;
                }
                
                for(String word : words){
                    if(canConvert(current, word)){
                        q.offer(word);
                    }
                }
            }
            answer++;
        }
        
        
        return 0;
    }
    
    public boolean canConvert(String str1, String str2){
        int count = 0; 
        for(int i = 0; i < str1.length(); i++){
            if(str1.charAt(i) != str2.charAt(i)){
                count++;
            }
        }
        return count == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    
    if target not in words : 
        return  0
    
    return bfs(begin, target, words)


def bfs(begin, target, words):
    queue = deque()
    queue.append([begin, 0])
    
    while queue:
        now, step = queue.popleft()
        if now == target:
            return step
        
        for word in words:
            cnt = 0
            for i in range(len(word)):
                if word[i] != now[i]:
                    cnt += 1
                    
            if cnt == 1:
                    queue.append([word, step+1])
        